### DISCUSSION TITLE
A build system for using C/C++ code

### DISCUSSION THREAD

**Discussion by lbalazscs** [lbalazscs/Pixelitor#282] [Ideas]
A lot of open-source code for image processing exists only as C/C++ libraries. To access C++ libraries, it may be worth using the [Foreign Function & Memory API](https://openjdk.java.net/jeps/424),  even though it's still in Preview state in Java 19.
Some libraries we could use:
- [ImageMagick](https://imagemagick.org/) - this is already used for import/export, but only through the command line, which is slower than necessary because all images are encoded in PNG before sending them to ImageMagick.
- [G'MIC](https://gmic.eu/) - it has hundreds of image filters, and an embedded interpreter for creating new filters. 
- [GEGL](https://gegl.org/) - many of Gimp's filters are implemented here.

It was a long time ago (20 years...) when I was last seriously involved in C/C++ programming, I don't really know what would be the best way to use such libraries, I don't know what is good, what are the best practices, tradeoffs. I'd appreciate ideas in this area. As far as I understand, the best approach would be to copy the C/C++ source files into this repository, and have some command that builds .so and .dll files from them. It would be nice if all of this could also run in GitHub Actions. 

(also see issue #235 the origins of this discussion)

**Comment by elfring** [comment]
>…, I don't know what is good,

I guess that you might know more about the application of some software libraries than you indicate by such a wording.


>…, the best approach would be to copy the C/C++ source files into this repository,

:thought_balloon: I suggest to reconsider this view in more detail for dependency management goals.


>and have some command that builds .so and .dll files from them.

:thinking: How do you think about to [take another look at build generation tools](https://en.wikipedia.org/wiki/List_of_build_automation_software#Build_script_generation "Build script generation")?


>It would be nice if all of this could also run in GitHub Actions.

For which actions would you become interested?

**Comment by 2dpdlja496or43iq** [comment]
As a Pixelitor user I'm strongly against this idea. I'm using Pixelitor only because it's a pure Java application (together with JEdit, MuCommander, NetBeans,...). Please keep Pixelitor a pure Java application!

**Comment by lbalazscs** [comment]
Well, this isn't going to happen for at least a year because even in Java 21, the Foreign Function & Memory API will still be in preview. Plus, it's going to be optional: the "core" (pure Java) Pixelitor will keep working even if you don't have or can't use the native libraries. Just like Pixelitor can use ImageMagick right now (through the command-line interface), but it doesn't *require* it.

**Comment by lbalazscs** [comment]
In the meantime, I realized that in the C++ world there is no such thing as a universally accepted best practice for project setup. Therefore, the original question can't be answered, and I am closing this discussion. 

As for package managers, Conan seems to be the most popular. However, it is useless for our purposes since it doesn't have gegl and libgmic.
